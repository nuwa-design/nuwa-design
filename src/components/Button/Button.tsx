/**
 * @FileName src/components/Button/Button.tsx
 * @Description 按钮组件 - Core
 * @Author qulongjun <contact@nuwa.design>
 * @Date 2021/11/21 21:45:48
 */
import * as React from 'react';
import propTypes from 'prop-types';
import classNames from 'classnames';
import { ICommonProps } from '@interfaces/ICommonProps';
import { getGlobalStyleConfig } from '../../utils';
import {
  ThemeColor,
  ElementSize,
  BtnShape,
  BtnType,
  BtnHtmlType,
  ElementSizeArr,
  ThemeColorArr,
  BtnTypeArr,
  BtnHtmlTypeArr,
  BtnShapeArr,
} from '../../types';
import { SizeMap } from '../../configs';
import './style/index.less';

const NUWA_PREFIX = getGlobalStyleConfig().nuwaPrefix;

export interface IButtonProps extends ICommonProps {
  /** 设置按钮样式类型 */
  type?: ThemeColor;
  /** 设置按钮悬停时样式类型 */
  hoverType?: ThemeColor;
  /** 设置按钮形状 */
  btnShape?: BtnShape;
  /** 设置按钮行为类型 */
  btnType?: BtnType;
  /** 设置按钮原生类型，可选值请参考 HTML 标准 */
  htmlType?: BtnHtmlType;
  /** 设置按钮大小 */
  size?: ElementSize;
  /** 是否显示为块状布局 */
  block?: boolean;
  /** 是否显示为禁用状态 */
  disabled?: boolean;
  /** 是否显示为激活状态 */
  active?: boolean;
  /** 是否显示为仅图标按钮 */
  iconOnly?: boolean;
  /** 按钮点击事件, event: 点击响应事件 */
  onClick?: (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;
}

const Button: React.FC<IButtonProps> = (props: IButtonProps) => {
  const {
    type,
    hoverType,
    btnType,
    btnShape,
    htmlType,
    size,
    block,
    disabled,
    iconOnly,
    active,
    className,
    children,
    onClick,
  } = props;

  const realSize = typeof size === 'string' ? SizeMap[size] : null;

  const innerCls: string = classNames(
    'nuwa_btn',
    {
      [`${NUWA_PREFIX}_btn-${btnType && btnType === 'outlined' ? `${btnType}-` : ``}${type}`]:
        type && ThemeColorArr.includes(type),
      [`${NUWA_PREFIX}_btn-${btnType !== 'basic' ? `${btnType}-` : ``}hover-${hoverType}`]:
        hoverType && ThemeColorArr.includes(hoverType),
      [`${NUWA_PREFIX}_btn-${btnShape}`]: btnShape && BtnShapeArr.includes(btnShape) && btnShape !== 'round',
      [`${NUWA_PREFIX}_btn-${realSize}`]: size && ElementSizeArr.includes(size) && size !== 'medium' && realSize,
      [`${NUWA_PREFIX}_btn-link`]: btnType === 'link',
      [`${NUWA_PREFIX}_btn-icon`]: iconOnly,
      [`${NUWA_PREFIX}-block`]: block,
      disabled: disabled && !active,
      active: active && !disabled,
    },
  );

  return (
    <button type={htmlType} className={classNames(innerCls, className)} onClick={onClick}>
      {children}
    </button>
  );
};

Button.propTypes = {
  type: propTypes.oneOf(ThemeColorArr),
  hoverType: propTypes.oneOf(ThemeColorArr),
  btnShape: propTypes.oneOf(BtnShapeArr),
  btnType: propTypes.oneOf(BtnTypeArr),
  htmlType: propTypes.oneOf(BtnHtmlTypeArr),
  size: propTypes.oneOf(ElementSizeArr),
  block: propTypes.bool,
  disabled: propTypes.bool,
  active: propTypes.bool,
  iconOnly: propTypes.bool,
  onClick: propTypes.func,
};

Button.defaultProps = {
  type: 'primary',
  btnType: 'basic',
  btnShape: 'round',
  htmlType: 'button',
  size: 'medium',
  block: false,
  disabled: false,
  iconOnly: false,
};

export default Button;
