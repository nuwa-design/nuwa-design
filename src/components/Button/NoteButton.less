.note-container {
  .note-bg {
    width: 100px;
    height: 100px;
    background-image: url('./css-avatar.png');
    background-size: contain;
  }
}
.note-button {
  padding: 0 18px;
  border-radius: 4px;
  background-color: #fefefe;
  display: flex;
  flex-direction: column;
  align-items: center;
}
