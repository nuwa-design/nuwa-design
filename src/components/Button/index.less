.button {
  padding: 0 18px;
  border-radius: 4px;
  background-color: #fefefe;
}

.button-status {
  color: #ffffff;
  text-align: center;
  padding: 10px;
  margin-top: 20px;
}
