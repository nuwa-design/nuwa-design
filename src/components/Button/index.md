---
nav:
  title: 组件
  order: 2
  path: /components
---

## Button

Demo:

```tsx
import React, { useCallback } from "react";
import { Button } from 'nuwa-design';

const onClick = useCallback((e) => {
  console.log('trigger click', e)
}, []);

export default () => <>
  <Button type="primary" onClick={onClick}>Primary</Button> &nbsp;
  <Button type="secondary">Secondary</Button>&nbsp;
  <Button type="success">Success</Button>&nbsp;
  <Button type="danger">Danger</Button>&nbsp;
  <Button type="info">Info</Button>&nbsp;
  <Button type="warning">Warning</Button><br /><br />
  <Button type="primary" btnType="outlined" onClick={onClick}>Primary</Button> &nbsp;
  <Button type="secondary" btnType="outlined">Secondary</Button>&nbsp;
  <Button type="success" btnType="outlined">Success</Button>&nbsp;
  <Button type="danger" btnType="outlined">Danger</Button>&nbsp;
  <Button type="info" btnType="outlined">Info</Button>&nbsp;
  <Button type="warning" btnType="outlined">Warning</Button>&nbsp;
  <br />
  <br />
  <Button type="primary" hoverType="danger" onClick={onClick}>Primary</Button> &nbsp;
  <Button type="secondary" hoverType="success">Secondary</Button>&nbsp;
  <Button type="success">Success</Button>&nbsp;
  <Button type="danger">Danger</Button>&nbsp;
  <Button type="info">Info</Button>&nbsp;
  <Button type="warning">Warning</Button><br /><br />
  <Button type="primary" hoverType="danger" btnType="outlined" onClick={onClick}>Primary</Button> &nbsp;
  <Button type="secondary" btnType="outlined" hoverType="warning">Secondary</Button>&nbsp;
  <Button type="success" btnType="outlined">Success</Button>&nbsp;
  <Button type="danger" btnType="outlined">Danger</Button>&nbsp;
  <Button type="info" btnType="outlined">Info</Button>&nbsp;
  <Button type="warning" btnType="outlined">Warning</Button>&nbsp;
  <br />
  <br />
  <Button type="primary" btnShape="round" onClick={onClick}>Primary</Button> &nbsp;
  <Button type="primary" btnShape="square" onClick={onClick}>Primary</Button> &nbsp;
  <Button type="primary" btnShape="pill" onClick={onClick}>Primary</Button> &nbsp;
  <Button type="primary" btnShape="circle" onClick={onClick}>1</Button> &nbsp;
  <Button type="primary" btnShape="round" btnType="outlined" onClick={onClick}>Primary</Button> &nbsp;
  <Button type="primary" btnShape="square" btnType="outlined" onClick={onClick}>Primary</Button> &nbsp;
  <Button type="primary" btnShape="pill" btnType="outlined" onClick={onClick}>Primary</Button> &nbsp;
  <Button type="primary" btnShape="circle" btnType="outlined" onClick={onClick}>1</Button> &nbsp;
</>;
```

<API></API>

[HTML 标准](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#attr-type)

