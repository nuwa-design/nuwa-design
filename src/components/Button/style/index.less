@import '../../../styles/index';

@btn-prefix-cls: ~'@{nuwa-prefix}-btn';


.@{btn-prefix-cls} {
  line-height: 30px;
}