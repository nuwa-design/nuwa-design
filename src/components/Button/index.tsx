/**
 * @FileName src/components/Button/index.tsx
 * @Description 按钮组件 - Entry
 * @Author qulongjun <contact@nuwa.design>
 * @Date 2021/11/21 19:35:41
 */
import Button from './Button';

export type { IButtonProps } from './Button';

export default Button;
