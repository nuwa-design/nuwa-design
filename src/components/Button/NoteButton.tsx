import * as React from 'react';
import Button from './Button';
import './NoteButton.less';

const NoteButton: React.FC = () => (
  <div className="note-container">
    <Button className="note-button">this is Note Button</Button>
    <div className="note-bg" />
  </div>
);

export default NoteButton;
