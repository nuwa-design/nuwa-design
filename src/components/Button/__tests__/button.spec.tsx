/**
 * @FileName src/components/Button/__test__/button.test.ts
 * @Description Button Component Test Cases
 * @Author qulongjun <contact@nuwa.design>
 * @Date 2021/11/20 16:01:57
 */
import React from 'react';
import { mount, render } from 'enzyme';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faTag } from '@fortawesome/free-solid-svg-icons';
import mountTest from '../../../../tests/mountTest';
import Button from '../Button';

describe('Button', () => {
  mountTest(Button);
  mountTest(() => <Button type='primary' />);
  mountTest(() => <Button btnType='outlined' />);
  mountTest(() => <Button hoverType='success' />);
  mountTest(() => <Button btnShape='square' />);
  mountTest(() => <Button htmlType='submit' />);
  mountTest(() => <Button size='small' />);
  mountTest(() => <Button block />);
  mountTest(() => <Button active />);
  mountTest(() => <Button disabled />);
  mountTest(() => <Button iconOnly />);
  mountTest(() => <Button customTag='a' />);
  mountTest(() => <Button customTag='div' />);

  it('renders correctly', () => {
    expect(<Button>Follow</Button>).toMatchSnapshot();
  });

  it('mount correctly', () => {
    expect(() => mount(<Button>Follow</Button>)).not.toThrow();
  });

  it('should render empty button without errors', () => {
    const wrapper = mount(
      <Button>
        {null}
        {undefined}
      </Button>,
    );
    expect(wrapper.render()).toMatchSnapshot();
  });

  it('should not clickable when button is disabled', () => {
    const onClick = jest.fn();
    const wrapper = mount(
      <Button disabled onClick={onClick}>
        button
      </Button>,
    );
    wrapper.simulate('click');
    expect(onClick).not.toHaveBeenCalledWith();
  });

  it('should not set error customTag', () => {
    const wrapper = mount(
      <Button disabled customTag={null as any}>
        button
      </Button>,
    );
    expect(wrapper.render()).toMatchSnapshot();
  });

  it('warns if size is wrong', () => {
    const mockWarn = jest.fn();
    jest.spyOn(console, 'error').mockImplementation(mockWarn);
    const size = 30;
    render(<Button size={size as any} />);
    expect(mockWarn).toHaveBeenCalledTimes(2);
    expect(mockWarn.mock.calls[1][0]).toMatch(`unreachable size:${size}`);
  });

  it('render icon correctly', () => {
    expect(<Button>Button</Button>).toMatchSnapshot();
    expect(
      <Button>
        <FontAwesomeIcon icon={faTag} />
        Button
      </Button>,
    ).toMatchSnapshot();
    expect(
      <Button>
        Button
        <FontAwesomeIcon icon={faTag} />
      </Button>,
    ).toMatchSnapshot();
    expect(
      <Button>
        <i className='fa fa-tag' />
        Button
      </Button>,
    ).toMatchSnapshot();
    expect(
      <Button>
        Button
        <i className='fa fa-tag' />
      </Button>,
    ).toMatchSnapshot();
    expect(
      <Button iconOnly>
        <i className='fa fa-tag' />
      </Button>,
    ).toMatchSnapshot();
    expect(
      <Button iconOnly>
        <FontAwesomeIcon icon={faTag} />
      </Button>,
    ).toMatchSnapshot();
  });
  it('should not insert space to button', () => {
    const wrapper1 = mount(<Button btnType='link'>按钮</Button>);
    expect(wrapper1.text()).toBe('按钮');
    const wrapper2 = mount(<Button btnType='outlined'>按钮</Button>);
    expect(wrapper2.text()).toBe('按钮');
    const wrapper3 = mount(<Button>按钮</Button>);
    expect(wrapper3.text()).toBe('按钮');
  });
  it('should handle fragment as children', () => {
    const wrapper = mount(
      <Button>
        <>text</>
      </Button>,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
