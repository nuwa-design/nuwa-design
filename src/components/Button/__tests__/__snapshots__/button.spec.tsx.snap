// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button render icon correctly 1`] = `
<Button
  block={false}
  btnShape="round"
  btnType="basic"
  customTag="button"
  disabled={false}
  htmlType="button"
  iconOnly={false}
  size="medium"
  type="primary"
>
  Button
</Button>
`;

exports[`Button render icon correctly 2`] = `
<Button
  block={false}
  btnShape="round"
  btnType="basic"
  customTag="button"
  disabled={false}
  htmlType="button"
  iconOnly={false}
  size="medium"
  type="primary"
>
  <FontAwesomeIcon
    border={false}
    className=""
    fixedWidth={false}
    flip={null}
    icon={
      Object {
        "icon": Array [
          512,
          512,
          Array [],
          "f02b",
          "M0 252.118V48C0 21.49 21.49 0 48 0h204.118a48 48 0 0 1 33.941 14.059l211.882 211.882c18.745 18.745 18.745 49.137 0 67.882L293.823 497.941c-18.745 18.745-49.137 18.745-67.882 0L14.059 286.059A48 48 0 0 1 0 252.118zM112 64c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48z",
        ],
        "iconName": "tag",
        "prefix": "fas",
      }
    }
    inverse={false}
    listItem={false}
    mask={null}
    pull={null}
    pulse={false}
    rotation={null}
    size={null}
    spin={false}
    swapOpacity={false}
    symbol={false}
    title=""
    transform={null}
  />
  Button
</Button>
`;

exports[`Button render icon correctly 3`] = `
<Button
  block={false}
  btnShape="round"
  btnType="basic"
  customTag="button"
  disabled={false}
  htmlType="button"
  iconOnly={false}
  size="medium"
  type="primary"
>
  Button
  <FontAwesomeIcon
    border={false}
    className=""
    fixedWidth={false}
    flip={null}
    icon={
      Object {
        "icon": Array [
          512,
          512,
          Array [],
          "f02b",
          "M0 252.118V48C0 21.49 21.49 0 48 0h204.118a48 48 0 0 1 33.941 14.059l211.882 211.882c18.745 18.745 18.745 49.137 0 67.882L293.823 497.941c-18.745 18.745-49.137 18.745-67.882 0L14.059 286.059A48 48 0 0 1 0 252.118zM112 64c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48z",
        ],
        "iconName": "tag",
        "prefix": "fas",
      }
    }
    inverse={false}
    listItem={false}
    mask={null}
    pull={null}
    pulse={false}
    rotation={null}
    size={null}
    spin={false}
    swapOpacity={false}
    symbol={false}
    title=""
    transform={null}
  />
</Button>
`;

exports[`Button render icon correctly 4`] = `
<Button
  block={false}
  btnShape="round"
  btnType="basic"
  customTag="button"
  disabled={false}
  htmlType="button"
  iconOnly={false}
  size="medium"
  type="primary"
>
  <i
    className="fa fa-tag"
  />
  Button
</Button>
`;

exports[`Button render icon correctly 5`] = `
<Button
  block={false}
  btnShape="round"
  btnType="basic"
  customTag="button"
  disabled={false}
  htmlType="button"
  iconOnly={false}
  size="medium"
  type="primary"
>
  Button
  <i
    className="fa fa-tag"
  />
</Button>
`;

exports[`Button render icon correctly 6`] = `
<Button
  block={false}
  btnShape="round"
  btnType="basic"
  customTag="button"
  disabled={false}
  htmlType="button"
  iconOnly={true}
  size="medium"
  type="primary"
>
  <i
    className="fa fa-tag"
  />
</Button>
`;

exports[`Button render icon correctly 7`] = `
<Button
  block={false}
  btnShape="round"
  btnType="basic"
  customTag="button"
  disabled={false}
  htmlType="button"
  iconOnly={true}
  size="medium"
  type="primary"
>
  <FontAwesomeIcon
    border={false}
    className=""
    fixedWidth={false}
    flip={null}
    icon={
      Object {
        "icon": Array [
          512,
          512,
          Array [],
          "f02b",
          "M0 252.118V48C0 21.49 21.49 0 48 0h204.118a48 48 0 0 1 33.941 14.059l211.882 211.882c18.745 18.745 18.745 49.137 0 67.882L293.823 497.941c-18.745 18.745-49.137 18.745-67.882 0L14.059 286.059A48 48 0 0 1 0 252.118zM112 64c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48z",
        ],
        "iconName": "tag",
        "prefix": "fas",
      }
    }
    inverse={false}
    listItem={false}
    mask={null}
    pull={null}
    pulse={false}
    rotation={null}
    size={null}
    spin={false}
    swapOpacity={false}
    symbol={false}
    title=""
    transform={null}
  />
</Button>
`;

exports[`Button renders correctly 1`] = `
<Button
  block={false}
  btnShape="round"
  btnType="basic"
  customTag="button"
  disabled={false}
  htmlType="button"
  iconOnly={false}
  size="medium"
  type="primary"
>
  Follow
</Button>
`;

exports[`Button should handle fragment as children 1`] = `
<Button
  block={false}
  btnShape="round"
  btnType="basic"
  customTag="button"
  disabled={false}
  htmlType="button"
  iconOnly={false}
  size="medium"
  type="primary"
>
  <button
    className="nuwa_btn nuwa_btn-primary"
    type="button"
  >
    text
  </button>
</Button>
`;

exports[`Button should not set error customTag 1`] = `null`;

exports[`Button should render empty button without errors 1`] = `
<button
  class="nuwa_btn nuwa_btn-primary"
  type="button"
/>
`;
